contracts/products/SoteriaCoverageProduct.sol

```diff
@@ -0,0 +1,681 @@
+     * @param coverLimit cover limit.
+    */
+    function _minRequiredAccountBalance(uint256 coverLimit) internal returns (uint256 minRequiredAccountBalance) {
+        minRequiredAccountBalance = ( _maxRateNum * _chargeCycle * coverLimit ) / _maxRateDenom;
+    }
+
+    /**
+     * @notice Use _beforeTokenTransfer hook from ERC721 standard to ensure Soteria policies are non-transferable, and only one can be minted per user
+     * @dev This hook is called on mint, transfer and burn
+     * @param from sending address.
+     * @param to receiving address.
+     * @param tokenId tokenId.
+    */
+    function _beforeTokenTransfer(address from, address to, uint256 tokenId) internal virtual override {
+        super._beforeTokenTransfer(from, to, tokenId);
+        require(from == address(0), "only minting permitted");
+    }
+
+    /**
+     * @notice Starts the **cooldown** period for the user.
+     */
+    function _startCooldown(address policyholder) internal {
+        _cooldownStart[policyholder] = block.timestamp;
+        emit CooldownStarted(policyholder, _cooldownStart[policyholder]);
+    }
+
+    /**
+     * @notice Abandons the **cooldown** period for the user.
+     * @dev Original name for this function from the deprecated Vault.sol was stopCooldown()
+     * @dev Renamed this to _exitCooldown because "reset" gives the impression that you are starting the cooldown again, and "stop" gives the impression you are stopping the cooldown for it to pick up again later from where you stopped it
+     * @dev I thought "exit" gives the imagery that you are exitting the process, and you will need to restart it manually later (which is more accurate, if user calls updateCoverLimit(), they then need to redo 'deactivatePolicy() => withdraw()' to get their entire funds back)
+     */
+    function _exitCooldown(address policyholder) internal {
+        _cooldownStart[policyholder] = 0;
+        emit CooldownStopped(policyholder);
+    }
+
+    /**
+     * @notice Determine if cooldown has passed for a policy holder
+     * @return True if cooldown has passed, false if not
+     */
+    function _hasCooldownPassed(address policyholder) internal returns (bool) {
+        if (_cooldownStart[policyholder] == 0) {return false;}
+        else {return block.timestamp >= _cooldownStart[policyholder] + _cooldownPeriod;}
+    }
+
+    /**
+     * @notice Gets the unique referral code for a user.
+     * @param user_ The user.
+     * @return referralCode_ The referral code.
+     */
+    function _getReferralCode(address user_) internal view returns (uint256 referralCode_) {
+        return uint256(uint160(user_));
+    }
+
+    /**
+     * @notice Internal function to process referral code
+     * @param policyholder_ Policy holder
+     * @param coverLimit_ Cover limit
+     * @param referralCode_ Referral code
+     */
+    function _processReferralCode(address policyholder_, uint256 coverLimit_, uint256 referralCode_) internal {
```

Comment:
We cannot grant rewardPoints based on coverLimit, because user will game the system: activate largest possible coverLimit -> get a bunch of points -> then reset to what you wanted. Therefore need to just give out a flat rate amount (say 0.01 ETH)

Can people get a “referral” by just copying someones address? Not even our policyholders? 

Perhaps we need some kind of a seed to make sure referral works properly (maybe similar to permit() sigs):
1. Only active policyholder with coverLimit != 0 can be a referrer
2. The code must grant set amount of points
3. Code must work only once for referee, but referrer can earn points with every use of his code
4. Can’t refer yourself

Reply:
Currently points 3 and 4 are implemented
- Can't refer to self
- Referee can only use code once, referrer has no limitations on how many times his/her code can be used

Will need to implement the following
- Give flat rate (0.01 ETH default, make it a state variable that can be changed) on successful referral, rather than % of coverLimit to avoid the user gaming the system
- Set restriction that `coverLimitOf[referrer] > 0`

Currently the user can enter any referral code, I was thinking a user can game this by using the referral code of another Ethereum address they own. Will need to think some more on how to mitigate this.

Reply:
Simplest way I can think of to mitigate a user gaming referrals by using a referral code of another address they own, is to increase the restriction to `coverLimitOf[referrer] > 1 ether`. This way yes, they can still use a referral code of another address they own, but they will still be charged a premium on the referring account.

If we want another protocol/party to have a referral code without them paying to open a Soteria account, we can open an account for them with activatePolicy(...) function

Reply:
The edge case where you refer one of your accounts to another is actually totally fine! As long as both address are actual policyholders, this is great. What I was concerned about is creating a referral code without the owner of that address: if I can just check on etherscan who's the policyholder and buy with their address, then it defeats the purpose of referral code completely, might as well just grant free points to every policyholder

Reply:
Ah ok

So basically we want to avoid the following: Any user can go to Etherscan, find a policyholder, get their referral code from the policyholder address, and enter the referral code.

This means we'll have to generate the referral code off-chain (to make it something that users can't get from just Etherscan view functions), and verify it on-chain. Most standard way to do this, off the top of my head, is permit() and ERC-712 signatures.